keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,538 @@
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, a Kubernetes Service accepts a list of ports to be exposed by it. It is possible to specify any number of ports, including the entire port range, by enlisting them in the service spec. This can be tedious if the service needs a large number of ports. This KEP proposes to add a new field to the ServicePort spec to allow exposing the entire port range( 1 to 65535). 
+
+## Motivation
+
+There are several applications like SIP apps or RTP which needs a lot of ports to run multiple calls or media streams. Currently the only option is to specify every port in the
+Service Spec. A request for port ranges in Services has been open in https://github.com/kubernetes/kubernetes/issues/23864. Implementing port ranges are challenging since iptables/ipvs do not support port ranges. Hence, the proposal to set a single field in order to expose the entire port range and implement the service clients and endpoints accordingly.
+
+### Goals
+
+* Allow users to optionally expose the entire port range via a Service (of Type LoadBalancer or ClusterIP).
```

Comment:
@uablrek ~ can you comment on the kproxy supportability for this ? re: ipvs/iptables/nftables/...

Reply:
That iptables/ipvs doesn't support port-ranges is not entirely true, but they can *not* re-map ports in a range, like with an offset. The ports are unchanged.
From ipvsadm(8) man page;
```
       -A, --add-service
              Add a virtual service. A service address is uniquely defined  by
              a triplet: IP address, port number, and protocol. Alternatively,
              a virtual service may be defined by a firewall-mark.
```
As you can see you must use `firewall-mark` to load-balance a range or an entire ip. The rule to set the fwmark is done by ip(6)tables and can specify port-ranges and protocol (or not).

Reply:
In this case a single fwmark can be used to cover all services.  If we allowed arbitrary ranges, we'd quickly run out of marks.  Unless I misunderstand how they are used?

Reply:
How about defining allPorts service as fwmark-service? That doesn't appear to make it a persistent service. iptables rules will set the packet mark. 

If we used the --tcp-service/--udp-service option with the persistent behavior, this might be desirable based on use-case? If the clients are going to use a random port to reach the service, it is probably ok to send packets to the same service port to the same backend?

Reply:
The fwmark is a single 32 bit int as far as IPVS is concerned - no bitmask.  So while there are a lot of them, it collides with anyone else who uses fwmark.  We'd need to allocate a DURABLE mark per-service and add iptables rules for them, then consume them in ipvs.  They have to survive kube-proxy restarts because ipvs is not atomic (like iptables-restore).  

The persistent feature, if I understand is 3-tuple not 5-tuple.  Maybe that is OK?  It seems surprising...

Reply:
true.. managing the mask can be messy, since we would also need to deal with conflicting mark values used by other components (Calico, other networking addons).

I missed that the persistent feature is 3-tuple, not 5.

Reply:
I think the port forwarding behavior might be do-able with ipvs using ipsets, but I would need to play around with it a bit to be sure.

I think a larger problem with IPVS could be regarding it's "bind-all" behavior (https://github.com/kubernetes/kubernetes/issues/72236), where a all ports Service will conflict with a real host port. Long standing issue that we need to address, but have not come up with a good path forward yet.

Reply:
Services of type `allPorts` don't have node ports. So it is primarily
1. `ClusterIP:port` which will be dnated to `pod:port` 
2. `ExternalIP:port` which will be dnated to `pod:port`

Reply:
loadbalancers with no nodeports? those are a thing with https://github.com/kubernetes/enhancements/issues/1864

Reply:
There should just be a few allPorts services in a cluster, say max 16, and if fwmark is used only for those it should be managable. It is also a per-node resource so kube-proxy can manage them locally, e.g re-read after re-start. See also https://github.com/kubernetes/enhancements/pull/2611#issuecomment-817108069.